httpcore5-h2/src/test/java/org/apache/hc/core5/http2/examples/Http2GreetingServer.java

```diff
@@ -0,0 +1,181 @@
+        protected AsyncRequestConsumer<Message<HttpRequest, HttpEntity>> supplyConsumer(HttpRequest request,
+                                                                                        final EntityDetails
+                                                                                                entityDetails,
+                                                                                        HttpContext context) {
+            AsyncEntityConsumer entityConsumer = new NoopEntityConsumer();
+
+            if (entityDetails != null) {
+                entityConsumer = new AbstractClassicEntityConsumer<HttpEntity>(1024 * 64, Executors.newCachedThreadPool()) {
+                    @Override
+                    protected HttpEntity consumeData(ContentType contentType, InputStream inputStream) {
+                        return new InputStreamEntity(inputStream, entityDetails.getContentLength(), contentType);
+                    }
+                };
+            }
+            return new BasicRequestConsumer<HttpEntity>(entityConsumer);
+
+        }
+
+        @Override
+        protected void handle(Message<HttpRequest, HttpEntity> requestMessage,
+                              AsyncServerRequestHandler.ResponseTrigger responseTrigger,
+                              HttpContext context) throws HttpException, IOException {
+
+            try {
+                final HttpCoreContext coreContext = HttpCoreContext.adapt(context);
+                final EndpointDetails endpoint = coreContext.getEndpointDetails();
+                final HttpRequest req = requestMessage.getHead();
+                final HttpEntity httpEntity = requestMessage.getBody();
+
+                // generic success response:
+                HttpResponse resp = new BasicHttpResponse(200);
+
+                // recording the request
+                System.out.println(String.format("[%s] %s %s %s", new Date(),
+                        endpoint.getRemoteAddress().toString(),
+                        req.getMethod(),
+                        req.getPath()));
+
+                // Request without an entity - GET/HEAD/DELETE
+                if (httpEntity == null) {
+                    responseTrigger.submitResponse(new BasicResponseProducer(resp), context);
+                    return;
+                }
+
+                // Request with an entity - POST/PUT
+                final ContentType entityContentType = ContentType.parse(httpEntity.getContentType());
+                String greeting = "Hello stranger\n";
+
+                if (entityContentType.getMimeType().equals(ContentType.APPLICATION_FORM_URLENCODED.getMimeType())) {
+
+                    Charset charset = ContentType.APPLICATION_FORM_URLENCODED.getCharset();
+                    if (entityContentType.getCharset() != null) {
+                        charset = entityContentType.getCharset();
+                    }
+
+                    StringBuilder contents = new StringBuilder();
```

Comment:
@rferreira This way of parsing request entity content is not very efficient or even very inefficient. It is confusing to see the request entity content represented as a blocking `InputStream` only to be read line by line into a `StringBuilder` and then converted to `String`. I imagine you had your reasons for doing so but this may not necessarily be a very good generic example.

Reply:
Yeah the rationale was that this was simpler for someone to grasp. I'll revise this with the String Entity Consumer and see how it looks.